src/sidebar/test/service-url-test.js

```diff
@@ -1,20 +1,166 @@
+    './util/url-util': { replaceURLParams: replaceURLParams },
+  });
+
+  var annotationUI = fakeAnnotationUI();
+
+  var features = { flagEnabled: sinon.stub().returns(true) };
+
+  var settings = {};
+
+  var store = {
+    links: sinon.stub().returns(linksPromise),
+  };
+
+  return {
+    annotationUI: annotationUI,
+    store: store,
+    serviceUrl: serviceUrlFactory(annotationUI, features, settings, store),
+    replaceURLParams: replaceURLParams,
+  };
+}
+
+describe('links', function () {
+  context('before the API response has been received', function() {
+    beforeEach(function() {
+      // Create a serviceUrl function with an unresolved Promise that will
+      // never be resolved - it never receives the links from store.links().
+      var parts = createServiceUrl(new Promise(function() {}));
+
+      this.serviceUrl = parts.serviceUrl;
+      this.store = parts.store;
+    });
+
+    it('sends one API request for the links at boot time', function() {
+      assert.isTrue(this.store.links.calledOnce);
+      assert.isTrue(this.store.links.calledWithExactly());
+    });
+
+    it('returns an empty string for any link', function() {
+      assert.equal(this.serviceUrl('foo'), '');
+    });
+
+    it('returns an empty string even if link params are given', function() {
+      assert.equal(this.serviceUrl('foo', {bar: 'bar'}), '');
+    });
   });
 
-  it('returns route URLs', function () {
-    assert.equal(service('help'), 'https://test.hypothes.is/docs/help');
+  context('if the API request fails', function() {
+    it('just keeps returning empty strings for URLs', function() {
+      var linksPromise = Promise.reject();
+
+      var serviceUrl = createServiceUrl(linksPromise).serviceUrl;
+
+      assert.equal(serviceUrl('second_link'), '');
+    });
   });
 
-  it('expands route parameters', function () {
-    assert.equal(service('user', {user: 'jim'}),
-      'https://test.hypothes.is/u/jim');
+  context('after the API response has been received', function() {
+    beforeEach(function() {
+      // The links Promise that store.links() will return.
+      this.linksPromise = Promise.resolve({
+        first_link: 'http://example.com/first_page/:foo',
+        second_link: 'http://example.com/second_page',
+      });
+
+      var parts = createServiceUrl(this.linksPromise);
+
+      this.annotationUI = parts.annotationUI;
+      this.serviceUrl = parts.serviceUrl;
+      this.replaceURLParams = parts.replaceURLParams;
+    });
+
+    it('updates annotationUI with the real links', function() {
+      var annotationUI = this.annotationUI;
+
+      return this.linksPromise.then(function(links) {
+        assert.deepEqual(annotationUI.getState(), {links: links});
+      });
+    });
+
+    it('calls replaceURLParams with the path and given params', function() {
+      var serviceUrl = this.serviceUrl;
+      var replaceURLParams = this.replaceURLParams;
+
+      return this.linksPromise.then(function() {
```

Comment:
> It'd be nice if there was a way to encapsulate this in the context() or beforeEach() so it doesn't have to be repeated in every test.

I wouldn't worry about it. It isn't that much code and I didn't have any problem understanding what was going on when reading it.

> It'd be even nicer if store didn't return promises, although I'm not sure exactly how that would work

`store` is the API client. It's fundamental job is to make API requests, which need to be async to avoid blocking the UI while the request is happening.